iModelCore/ECPresentation/Source/Shared/Queries/QueryBuilderHelpers.cpp

```diff
@@ -963,48 +1000,36 @@ PresentationQueryContractFieldPtr QueryBuilderHelpers::CreateDisplayLabelField(U
-    RefCountedPtr<PresentationQueryContractSimpleField> defaultPropertyValueField = PresentationQueryContractSimpleField::Create(nullptr, labelClause);
+    Utf8String labelClause;
+    if (labelProperty)
+        labelClause = Utf8PrintfString("[%s].[%s]", selectClass.GetAlias().c_str(), labelProperty->GetName().c_str());
+    else
+        labelClause = "''";
 
-    PresentationQueryContractFieldPtr labelField = PresentationQueryContractFunctionField::Create(name, FUNCTION_NAME_GetECInstanceDisplayLabel,
-        {
-        classIdField,
-        instanceIdField,
-        defaultPropertyValueField,
-        PresentationQueryContract::CreateRelatedInstanceInfoField(relatedInstancePaths),
-        });
+    auto defaultPropertyValueField = PresentationQueryContractSimpleField::Create(nullptr, {std::move(labelClause)}, false);
 
-    if (!labelOverrideValueSpecs.empty())
+    auto labelField = ApplyLabelOverrides([&]
         {
-        auto fallback = labelField;
-
-        InstanceLabelOverrideSelectFieldsBuilder builder(schemaHelper, selectClass, *instanceIdField, *classIdField, labelRequestsStack);
-        for (InstanceLabelOverrideValueSpecification const* spec : labelOverrideValueSpecs)
-            {
-            spec->Accept(builder);
-            if (builder.DefinitelyHasValue())
-                break;
-            }
-
-        if (builder.GetSelectFields().empty())
-            return labelField;
-
-        if (builder.GetSelectFields().size() == 1 && builder.DefinitelyHasValue())
+        return PresentationQueryContractFunctionField::Create(name, FUNCTION_NAME_GetECInstanceDisplayLabel,
             {
-            labelField = builder.GetSelectFields().front();
-            labelField->SetName(name);
-            return labelField;
-            }
-
-        RefCountedPtr<PresentationQueryContractFunctionField> coalesceField = PresentationQueryContractFunctionField::Create(name,
-            "COALESCE", ContainerHelpers::TransformContainer<bvector<PresentationQueryContractFieldCPtr>>(builder.GetSelectFields()));
-
-        if (!builder.DefinitelyHasValue())
-            coalesceField->GetFunctionParameters().push_back(labelField);
-
-        labelField = coalesceField;
-        }
-
+            classIdField,
+            instanceIdField,
+            defaultPropertyValueField,
+            PresentationQueryContract::CreateRelatedInstanceInfoField(relatedInstancePaths),
+            });
+        }, name, schemaHelper, selectClass, classIdField, instanceIdField, labelOverrideValueSpecs, labelRequestsStack);
+
+    // If class ID select clause results in NULL (i.e. when class ID is taken from a joined table with no overlaps), then label query might fail.
+    // NOTE: It's assumed that class ID field and instance ID fields are non-null constants or they're properties of the same instance that will both be defined or NULL.
+    // This should result into the following clause:
+    // IIF(classId IS NULL, CAST(NULL AS TEXT), ...)

```

Comment:
I don't understand the described situation - could you provide the full ECSQL as an example?

Reply:
```sql
SELECT 1 AS [/ContractId/],
       GetInstanceKey(IFNULL([this].[ECClassId], 0), [this].[ECInstanceId]) AS [/ECInstanceKeys/],
       CAST(NULL AS TEXT) AS [/InputECInstanceKeys/],
       '[]' AS [/RelatedInstanceInfo/],
       GetInstanceKey(IFNULL([RelatedClass_1 ].[ECClassId], 0),
       [RelatedClass_1].[ECInstanceId]) AS [/key/RelatedClass_1],      
       GetECInstanceDisplayLabel(([RelatedClass_1].[ECClassId]), ([RelatedClass_1].[ECInstanceId]), '', '[]')) AS [/DisplayLabel/RelatedClass_1],

FROM bis.Element [this]
LEFT JOIN RelatedClass_1 ON ...
```

When computing value for `/DisplayLabel/RelatedClass_1`, it calls `GetECInstanceDisplayLabel`, which throws an error if 1 or 2 argument is null.

Reply:
maybe the `IIF(classId IS NULL, ...` should be placed on the call to `GetECInstanceDisplayLabel` function call then, rathen than the whole label field?

Reply:
DisplayLabel clause may consist of `COALESCE` with many fallback specifications. For example:
`COALSESCE(GetDisplayLabel(classId, instanceId, ...), [RelatedClass_1].[Name], [RelatedClass_1].[UserLabel], ...)`
I think, it is easier (and probably more efficient) not to invoke `COALESCE` at all if there are no related instances found than to ensure that each fallback specification can work with NULL.

Reply:
In theory, some of the fallbacks in `COALESCE` could be static and not require the instance to exist to have a value. Now that I think of it, maybe the best approach would be to modify the `GetECInstanceDisplayLabel` custom function to return `NULL` if class or instance IDs are `NULL`?

Reply:
Made it so now `ContentQueryBuilder` wraps the column with `IIF` and not the `QueryBuilderHelpers::CreateDisplayLabelField`. As a result, hierarchy node hashes are not affected anymore.
Also extracted related instance display label field creation to `QueryBuilderHelpers`, to avoid repetition in `ContentQueryBuilder` tests.